File: monolith/backend/Dto/TemplateRenderModels.cs
using System.Text.Json.Serialization;

namespace CV_Generator.Dto;

public class TemplateRenderRequest
{
    [JsonPropertyName("user_id")]
    public Guid UserId { get; set; }

    [JsonPropertyName("extraction_id")]
    public Guid ExtractionId { get; set; }

    [JsonPropertyName("template_id")]
    public string? TemplateId { get; set; }

    [JsonPropertyName("language")]
    public string? Language { get; set; }

    [JsonPropertyName("tone")]
    public string? Tone { get; set; }

    [JsonPropertyName("save_to_documents")]
    public bool SaveToDocuments { get; set; } = true;

    [JsonPropertyName("title")]
    public string? Title { get; set; }
}

public class TemplateRenderSubmitResponse
{
    [JsonPropertyName("run_id")]
    public Guid RunId { get; set; }
}

public class TemplateRenderStatusResponse
{
    [JsonPropertyName("run_id")]
    public Guid RunId { get; set; }

    [JsonPropertyName("status")]
    public string Status { get; set; } = string.Empty;

    [JsonPropertyName("current_step")]
    public int CurrentStep { get; set; }

    [JsonPropertyName("steps")]
    public List<StepStatusDto> Steps { get; set; } = new();

    [JsonPropertyName("error_message")]
    public string? ErrorMessage { get; set; }

    [JsonPropertyName("created_at")]
    public DateTime CreatedAt { get; set; }

    [JsonPropertyName("completed_at")]
    public DateTime? CompletedAt { get; set; }

    [JsonPropertyName("cancelled_at")]
    public DateTime? CancelledAt { get; set; }
}

public class TemplateRenderResultResponse
{
    [JsonPropertyName("run_id")]
    public Guid RunId { get; set; }

    [JsonPropertyName("extraction")]
    public object? Extraction { get; set; }

    [JsonPropertyName("search")]
    public object? Search { get; set; }

    [JsonPropertyName("render")]
    public object? Render { get; set; }

    [JsonPropertyName("tex")]
    public string Tex { get; set; } = string.Empty;

    [JsonPropertyName("pdf_url")]
    public string? PdfUrl { get; set; }

    [JsonPropertyName("cv_id")]
    public Guid? CvId { get; set; }

    [JsonPropertyName("cv_version_id")]
    public Guid? CvVersionId { get; set; }

    [JsonPropertyName("saved")]
    public bool Saved { get; set; }

    [JsonPropertyName("match_score")]
    public double? MatchScore { get; set; }

    [JsonPropertyName("gap_skills")]
    public List<string> GapSkills { get; set; } = new();
}

// Server-Sent Events pushed by GET /events while a run executes.
public class TemplateRenderSseEvent
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("status")]
    public string? Status { get; set; }

    [JsonPropertyName("current_step")]
    public int? CurrentStep { get; set; }

    [JsonPropertyName("steps")]
    public List<StepStatusDto>? Steps { get; set; }

    [JsonPropertyName("error_message")]
    public string? ErrorMessage { get; set; }

    [JsonPropertyName("result")]
    public TemplateRenderResultResponse? Result { get; set; }
}

File: frontend/src/app/models/template-render.models.ts
import { StepStatus } from './cv-generation.models';

export type TemplateRenderRunStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

export interface TemplateRenderStatus {
  run_id: string;
  status: TemplateRenderRunStatus;
  current_step: number;
  steps: StepStatus[];
  error_message: string | null;
  created_at: string;
  completed_at: string | null;
  cancelled_at: string | null;
}

export interface TemplateRenderResult {
  run_id: string;
  extraction: any;
  search: any;
  render: any;
  tex: string;
  pdf_url: string | null;
  cv_id: string | null;
  cv_version_id: string | null;
  saved: boolean;
  match_score: number | null;
  gap_skills: string[];
}

export interface TemplateRenderSubmitResponse {
  run_id: string;
}

export interface TemplateRenderSseSnapshot {
  type: 'snapshot';
  status: TemplateRenderRunStatus;
  current_step: number;
  steps: StepStatus[];
  error_message?: string | null;
}

export interface TemplateRenderSseDone {
  type: 'done';
  status: TemplateRenderRunStatus;
  steps: StepStatus[];
  error_message?: string | null;
  result?: TemplateRenderResult;
}

export interface TemplateRenderSseError {
  type: 'error';
  error_message?: string;
}

export type TemplateRenderSseEvent = TemplateRenderSseSnapshot | TemplateRenderSseDone | TemplateRenderSseError;

// The 4 rows shown in the page's step list. Step 0 (Job Extraction) is
// satisfied before the run starts (fresh extraction or a picked history run);
// steps 1-3 map 1:1 to the backend run steps.
export const TEMPLATE_RENDER_RUN_STEPS: ReadonlyArray<{ name: string; runStep?: number }> = [
  { name: 'Job Extraction' },
  { name: 'Profile Matching', runStep: 0 },
  { name: 'Template Rendering', runStep: 1 },
  { name: 'PDF & Save', runStep: 2 },
];